torch/csrc/distributed/rpc/tensorpipe_agent.cpp

```diff
@@ -110,6 +113,48 @@ std::vector<c10::Stream> getCurrentStreamsForDevices(
   return streams;
 }
 
+std::vector<c10::Device> getDevicesOfTensors(
+    const std::vector<torch::Tensor>& tensors) {
+  c10::optional<c10::impl::VirtualGuardImpl> impl;
+  size_t deviceCount = 0;
+  std::vector<bool> indexBitSet;
+  for (const torch::Tensor& tensor : tensors) {
+    if (!tensor.is_cpu()) {
+      c10::Device device = tensor.device();
+      if (!impl.has_value()) {
+        impl.emplace(device.type());
+        indexBitSet.resize(impl->deviceCount());
+      }
+      TORCH_INTERNAL_ASSERT(device.type() == impl->type());
+      TORCH_INTERNAL_ASSERT(device.has_index());
+      if (!indexBitSet[device.index()]) {
+        deviceCount++;
+        indexBitSet[device.index()] = true;
+      }
+    }
+  }
+  std::vector<c10::Device> devices;
+  devices.reserve(deviceCount);
+  for (c10::DeviceIndex idx = 0; idx < indexBitSet.size(); idx++) {
```

Comment:
lint failures look real.

Reply:
The linter is correct "in principle" but here we know that the size of the vector will fit in a `DeviceIndex` because we used a `DeviceIndex` to resize the vector.

I'm starting to think that perhaps we should create a dedicated DeviceVector class, that can provide properly-typed accessors, can perhaps use small-vector optimizations, ...